### ISSUE TITLE
WRN_UnreferencedVarAssg "The variable '...' is assigned but its value is never used" is not reported for pattern variables

### ISSUE THREAD

**Issue by AlekseyTs** [dotnet/roslyn#9021]
```
    void Test3()
    {
        let x3 = 12;
        var x4 = 12;
    }
```

Observed:

```
    // (11,13): warning CS0219: The variable 'x4' is assigned but its value is never used
    //         var x4 = 12;
    Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "x4").WithArguments("x4").WithLocation(11, 13)
```

Expected:
WRN_UnreferencedVarAssg should also be reported for x3.

**Comment by gafter** [comment]
The `let` statement is not part of the pattern-matching feature at this time.

Pattern variables are frequently introduced as placeholders for values that are discarded, particularly when using the type pattern. It would therefore be inappropriate to warn that they are unused.

**Comment by AlekseyTs** [comment]
> It would therefore be inappropriate to warn that they are unused.

Is this a decision explicitly made by an LDM? Could you give a link to the notes?

**Comment by gafter** [comment]
This category of diagnostic is not part of the language spec.

**Comment by AlekseyTs** [comment]
> This category of diagnostic is not part of the language spec.

It feels like the decision should be made by some sort of design committee, even if it is not an LDM. What committee made the decision "It would therefore be inappropriate to warn that they are unused."? Could you give a link to the notes?
If you made this decision by yourself, please bring this issue to LDM. It is still appropriate to discuss and make decisions about categories that aren't part of the language spec at the LDM. If you can suggest a more suitable design committee, that would work too.

**Comment by gafter** [comment]
@AlekseyTs The 6/24 minutes are not published yet, but tentatively made the identifier required in a type pattern to resolve a number of ambiguous situations, including the issues around tuple patterns. This (not warning when the identifier is not used) is a natural consequence of that.

**Comment by AlekseyTs** [comment]
> This (not warning when the identifier is not used) is a natural consequence of that.

LDM should explicitly acknowledge that. Please bring this to the next LDM.

**Comment by AlekseyTs** [comment]
Reopening pending explicit decision of design committee on the issue.

**Comment by jaredpar** [comment]
Looks like a pattern design issue, reassigning to @gafter

**Comment by gafter** [comment]
I already resolved it. I will do so again.

**Comment by AlekseyTs** [comment]
Here is a unit-test that doesn't use 'let', I am going to check it in it shortly.
```
        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/9021")]
        [WorkItem(9021, "https://github.com/dotnet/roslyn/issues/9021")]
        public void WRN_UnreferencedVarAssg_01()
        {
            var source =
@"
public class X
{
    public static void Main()
    {
        var x = 11;
        Dummy(11 is var y);
    }

    static bool Dummy(params object[] x) {return true;}
}
";
            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: patternParseOptions);
            compilation.VerifyDiagnostics(
                // (6,13): warning CS0219: The variable 'x' is assigned but its value is never used
                //         var x = 11;
                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "x").WithArguments("x").WithLocation(6, 13),
                // (7,25): warning CS0219: The variable 'y' is assigned but its value is never used
                //         Dummy(11 is var y);
                Diagnostic(ErrorCode.WRN_UnreferencedVarAssg, "y").WithArguments("y").WithLocation(6, 25)
                );
        }
```

**Comment by gafter** [comment]
The LDM has confirmed - none of the members have asked to change the status quo that there are no diagnostics for these situations.

**Comment by gafter** [comment]
Assigning to @AlekseyTs to ensure that we add no unit tests that check this situation contrary to spec.

**Comment by gafter** [comment]
Due to [changes recently proposed by the LDM](https://github.com/dotnet/roslyn/issues/14794), if we implement wildcards in C#7 we will likely want this warning.

**Comment by gafter** [comment]
The LDM met yesterday and confirmed that we do not intend there to be new warnings for these scenarios, though IDEs and other tools may want to produce helpful style guidance.

**Comment by jcouv** [comment]
For the record, the [pertinent LDM notes](https://github.com/dotnet/csharplang/blob/master/meetings/2016/LDM-2016-11-30.md#warn-about-unused-expression-variables) (no "unused" warning on declaration expressions, including deconstructions, etc).
